VideoLocker/src/main/java/org/edx/mobile/http/OauthRefreshTokenAuthenticator.java

```diff
@@ -0,0 +1,76 @@
+package org.edx.mobile.http;
+
+import android.content.Context;
+import android.support.annotation.NonNull;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.inject.Inject;
+import com.jakewharton.retrofit.Ok3Client;
+
+import org.edx.mobile.authentication.LoginService;
+
+import org.edx.mobile.authentication.AuthResponse;
+import org.edx.mobile.logger.Logger;
+import org.edx.mobile.module.prefs.PrefManager;
+import org.edx.mobile.util.Config;
+
+import java.io.IOException;
+
+import okhttp3.Authenticator;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.Route;
+import retrofit.RestAdapter;
+import roboguice.RoboGuice;
+
+/**
+ * Created by cleeedx on 4/8/16.
+ */
+public class OauthRefreshTokenAuthenticator implements Authenticator {
+
+    private final Logger logger = new Logger(getClass().getName());
+    private Context context;
+
+
+    @Inject
+    Config config;
+
+    public OauthRefreshTokenAuthenticator(Context context) {
+        this.context = context;
+        RoboGuice.injectMembers(context, this);
+    }
+
+    @Override
+    public Request authenticate(Route route, final Response response) throws IOException {
+        logger.warn(response.toString());
+
+        if (!response.peekBody(200).string().contains("token_expired")) {
+            // Only get refresh token when we receive the "token_expired" message
+            // This check could be better since it depends on the server returning this message.
+            return null;
+        }
+
+        PrefManager pref = new PrefManager(context, PrefManager.Pref.LOGIN);
+        refreshAccessToken(pref);
+        return response.request().newBuilder()
+                .header("Authorization", pref.getCurrentAuth().token_type + " " + pref.getCurrentAuth().access_token)
+                .build();
+    }
+
+    private void refreshAccessToken(@NonNull PrefManager pref) {
+        OkHttpClient client = OkHttpUtil.getClient(context);
+        RestAdapter restAdapter = new RestAdapter.Builder()
+                .setClient(new Ok3Client(client))
+                .setEndpoint(config.getApiHostURL())
+                .build();
+        LoginService loginService = restAdapter.create(LoginService.class);
```

Comment:
The `OkHttpClient` could be passed as a parameter. Even the Retrofit `RestAdapter` could be injected as a `Provider`. However, per the current implementation, we would still need to generate the `LoginAPI` (or some other API interface) instance, if we want to use Retrofit to make the refresh call. This could be eliminated by using a `Provider` for it too, possibly via [`@AssistedInject`](https://github.com/google/guice/wiki/AssistedInject).

See https://github.com/google/guice/wiki/CyclicDependencies for more information on this.